runtime/ops/tls.rs

```diff
@@ -70,6 +85,421 @@ impl StoresClientSessions for ClientSessionMemoryCache {
+  }
+
+  fn new_server(tcp: TcpStream, tls_config: &Arc<ServerConfig>) -> Self {
+    let tls = TlsSession::Server(ServerSession::new(tls_config));
+    Self::new(tcp, tls)
+  }
+
+  pub fn into_split(self) -> (TlsStreamReader, TlsStreamWriter) {
+    let shared = Shared::new(self);
+    let rd = TlsStreamReader {
+      shared: shared.clone(),
+    };
+    let wr = TlsStreamWriter { shared };
+    (rd, wr)
+  }
+
+  fn poll_io(
+    &mut self,
+    cx: &mut Context<'_>,
+    for_reading: bool,
+    for_writing: bool,
+  ) -> Poll<io::Result<()>> {
+    assert!((for_reading && !for_writing) || (!for_reading && for_writing));
+
+    // Initially, assume that the TCP stream is both readable and writable. If
+    // this turns out not to be the case, `read_tls` or `write_tls()` will fail
+    // with `WouldBlock`; only then do we arrange to be asynchronously notified
+    // of read/write.
+    let mut tcp_read_ready = true;
+    let mut tcp_write_ready = true;
+
+    loop {
+      while tcp_write_ready && self.tls.wants_write() {
+        let mut wrapper = ImplementWriteTrait(&mut self.tcp);
+        match self.tls.write_tls(&mut wrapper) {
+          Ok(_) => {}
+          Err(err) if err.kind() == ErrorKind::WouldBlock => {
+            match self.tcp.poll_write_ready(cx) {
+              Poll::Ready(result) => result?,
+              Poll::Pending => tcp_write_ready = false,
+            }
+          }
+          Err(err) => return Poll::Ready(Err(err)),
+        }
+      }
+
+      while tcp_read_ready
+        && self.tls.wants_read()
+        && self.rd_shut != Shut::TcpShut
+      {
+        let mut wrapper = ImplementReadTrait(&mut self.tcp);
+        match self.tls.read_tls(&mut wrapper) {
+          Ok(0) if self.tls.is_handshaking() => {
+            let err =
+              io::Error::new(ErrorKind::UnexpectedEof, "tls handshake eof");
+            return Poll::Ready(Err(err));
+          }
+          Ok(0) => self.rd_shut = Shut::TcpShut,
+          Ok(_) => match self.tls.process_new_packets() {
+            Ok(_) => {}
+            Err(err) => {
+              let err = io::Error::new(ErrorKind::InvalidData, err);
+              return Poll::Ready(Err(err));
+            }
+          },
+          Err(err) if err.kind() == ErrorKind::WouldBlock => {
+            match self.tcp.poll_read_ready(cx) {
+              Poll::Ready(result) => result?,
+              Poll::Pending => tcp_read_ready = false,
+            }
+          }
+          Err(err) => return Poll::Ready(Err(err)),
+        }
+      }
+
+      if tcp_write_ready && self.tls.wants_write() {
+        continue;
+      }
+
+      if for_reading
+        && !self.tls.is_handshaking()
+        && (!self.tls.wants_read() || self.rd_shut == Shut::TcpShut)
+      {
+        return Poll::Ready(Ok(()));
+      }
+      if for_writing && !self.tls.is_handshaking() && !self.tls.wants_write() {
+        return Poll::Ready(Ok(()));
+      }
+
+      return Poll::Pending;
+    }
```

Comment:
It won't starve because:
* `wants_read()` will return `false` when there rustls has decrypted plaintext waiting for the user to pick up.
  No plaintext is read in this loop, so there's an upper limit to the amount of reading that happens each round.
* `wants_write()` returns true only as long as it has unsent, encrypted ciphertext in its internal buffer.
  No plaintext is written in this loop, so the upper bound for writes is defined by plaintext that the user had already sent before we got here.

Reply:
https://github.com/ctz/rustls/blob/b9a5a061b48598545168da41f759c2a191ad4748/rustls/src/server/mod.rs#L548-L556

Reply:
Okay, that makes sense but I don't think it's obvious to a casual (or even not-so-casual) reader. Can you add a comment or two?

Reply:
Done!